### PR TITLE
Put filters in an array instead of flt strings

### PR DESCRIPTION
**Pull request by vera-liu**
Done:
 - get rid of flt_ strings, migrate old slice flt params to list of filters, backend filters are lists, frontend filters are arrays
 - delete empty filters in old slices during migration
 - make having_filters and filters separate fields
 - replace old input in filters with freeform react-select (having filters still use input)

After:
screen shot 2017-02-01 at 10 18 09 am
screen shot 2017-02-01 at 10 18 33 am
screen shot 2017-02-01 at 10 19 03 am

needs-review @mistercrunch

### CODE REVIEW

File: superset/__init__.py [hunk 1]
```diff
@@ -39,6 +40,41 @@ def cast_form_data(form_data):
         d[k] = v
     return d
 
+def cast_filter_data(form_data):
```
Comment (mistercrunch) [inline superset/__init__.py]: I think this function should be defined in the migration script as there will be no use for it beyond the migration.

File: superset/assets/javascripts/explorev2/stores/fields.js [hunk 2]
```diff
@@ -1140,6 +1140,20 @@ export const fields = {
     mapStateToProps: (state) => ({
       choices: (state.datasource) ? state.datasource.filterable_cols : [],
       datasource: state.datasource,
+      opChoices: ['in', 'not in'],
```
Comment (mistercrunch) [inline superset/assets/javascripts/explorev2/stores/fields.js]: I'm not sure if there's a good reason to have different `opChoices` for WHERE and HAVING, my feeling is that we could hard code a superset :) of both arrays in the Filter element itself.

File: superset/viz.py [hunk 3]
```diff
@@ -144,15 +144,16 @@ def query_filters(self, is_having_filter=False):
         """Processes the filters for the query"""
         form_data = self.form_data
         # Building filters
+        flts = []
```
Comment (mistercrunch) [inline superset/viz.py]: Now that we have comprehensive filters, I would just pass them "as is" to the `Datasource.query` methods (druid and sqla). We may need to alter the code a bit there though. That means there's probably no more need for this `query_filters` method anymore.

File: superset/assets/javascripts/explorev2/components/Filter.jsx [hunk 4]
```diff
@@ -61,24 +61,27 @@ export default class Filter extends React.Component {
       if (!filter.choices) {
         this.fetchFilterValues(filter.col);
       }
+    }
+    if (this.props.opChoices.indexOf('==') !== -1) {
```
Comment (mistercrunch) [inline superset/assets/javascripts/explorev2/components/Filter.jsx]: shouldn't this be based on `filter.op`?

**Review by mistercrunch — CHANGES_REQUESTED** [review]

**Comment by mistercrunch** [comment]
Oh also I was wondering whether we quote/not quote based on the target column?

File: superset/assets/javascripts/explorev2/stores/fields.js (hunk 2 above)
Comment (vera-liu) [inline superset/assets/javascripts/explorev2/stores/fields.js]: do you mean normal filters and having filters?

File: superset/assets/version_info.json [hunk 5]
```diff
@@ -1 +0,0 @@
-{"GIT_SHA": "2d08e240285288b71df98747ddd4b6cca3220c5a", "version": "0.15.2"}
```
Comment (ascott) [inline superset/assets/version_info.json]: did you mean to delete this line?

Comment (mistercrunch) [inline superset/assets/version_info.json]: this should be in the gitignore

File: superset/viz.py [hunk 6]
```diff
@@ -144,15 +144,16 @@ def query_filters(self, is_having_filter=False):
...
+        if is_having_filter and 'having_filters' in form_data:
+            flts = form_data['having_filters']
```
Comment (ascott) [inline superset/viz.py]: it's not clear to me the difference between `flts` and `filters` unless i dig in and read more of the code. would be nice to make these var names more descriptive.

**Comment by vera-liu** [comment]
@mistercrunch One thing that's not working is that React-select will create an option once ',' is typed, so we can't have filter values containing comma...still haven't found a way to get around this

**Comment by mistercrunch** [comment]
@vera-liu that's not a big deal to me, I'd like to move forward with this regardless

**Comment by mistercrunch** [comment]
Merging into my branch now

**Review by mistercrunch — APPROVED** [review]

**Pull request merged**